### PR TITLE
lite ipld and decoder

### PR DESCRIPTION
**Pull request by turuslan**
- `Ipld2`: cbor-blake only, throw rare/unexpected (corruption, out of space) errors, reuse buffer.
- constexpr actor codes.
- lite decoders: address, cid, cbor, actor, h/amt.

### CODE REVIEW

File: core/storage/leveldb/prefix.cpp [hunk 1]
```diff
@@ -151,4 +151,8 @@ namespace fc::storage {
   void OneKey::set(Buffer value) {
     map->put(key, std::move(value)).value();
   }
+
+  void OneKey::remove() {
+    map->remove(key).value();
```
Comment (ortyomka) [inline core/storage/leveldb/prefix.cpp:156]: why do you get value() here?

File: test/core/vm/state/state_tree_test.cpp [hunk 2]
```diff
@@ -73,3 +78,27 @@ TEST_F(StateTreeTest, RegisterNewAddressLookupId) {
   EXPECT_OUTCOME_EQ(tree->registerNewAddress(address), kAddressId);
   EXPECT_OUTCOME_EQ(tree->lookupId(address), kAddressId);
 }
+
+TEST_F(StateTreeTest, Walk) {
```
Comment (ortyomka) [inline test/core/vm/state/state_tree_test.cpp:82]: looks hard, add annotation please

File: test/testutil/storage/ipld2.hpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        return false;
+      }
+      if (value) {
+        *value = std::move(r.value());
+      }
+      return true;
+    }
+
+    void put(const Hash256 &key, BytesIn value) override {
+      ipld->set(asCborBlakeCid(key), Buffer{value}).value();
```
Comment (ortyomka) [inline test/testutil/storage/ipld2.hpp:29]: again `value()`

**Review by ortyomka — APPROVED** [review]

File: core/storage/leveldb/prefix.cpp (hunk 1 above)
Comment (turuslan) [reply core/storage/leveldb/prefix.cpp:156]: Why can it fail? How offen will it fail? What can calling code do when it fails?

File: core/storage/leveldb/prefix.hpp [hunk 4]
```diff
@@ -60,11 +60,13 @@ namespace fc::storage {
   };
 
   struct OneKey {
+    OneKey() = default;
```
[the remark this answers is not in the document]
Comment (turuslan) [reply core/storage/leveldb/prefix.hpp:63]: Yes, it's expected.
To make it default constructible.

File: core/codec/cbor.hpp [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+      MAP,
+      CID,
+      SPECIAL,
+      INVALID,
+    };
+
+    Type type{Type::INVALID};
+    uint64_t extra{};
+
+    constexpr operator bool() const {
```
[the remark this answers is not in the document]
Comment (turuslan) [reply core/codec/cbor.hpp]: No, `constexpr` prevents dynamic memory allocations and such operations.
https://coliru.stacked-crooked.com/a/0012aa80af11b024

File: core/codec/amt.hpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+namespace fc::codec::amt {
+  struct AmtWalk {
+    Ipld2Ptr ipld;
+    std::vector<Hash256> cids;
+    size_t next_cid{};
+    Buffer _node;
+    BytesIn node;
+    size_t _values{};
+
+    inline AmtWalk(Ipld2Ptr ipld, Hash256 root) : ipld{ipld} {
```
[the remark this answers is not in the document]
Comment (turuslan) [reply core/codec/amt.hpp]: Thanks

File: core/storage/leveldb/prefix.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Alexey-N-Chernyshov) [reply core/storage/leveldb/prefix.hpp:63]: But why it must be default constructible? It is not used.
I prefer to forbid incorrect states.

File: test/core/storage/amt/amt_test.cpp [hunk 7]
```diff
@@ -149,3 +152,22 @@ TEST_F(AmtVisitTest, VisitError) {
                          return AmtError::kIndexTooBig;
                        }));
 }
+
+/*
+ * walk visits amt values
+ */
+TEST_F(AmtVisitTest, Walk) {
+  using namespace fc;
```
Comment (ortyomka) [inline test/core/storage/amt/amt_test.cpp:161]: Maybe wrap to `namespace fc {}`

File: test/core/vm/state/state_tree_test.cpp [hunk 8]
```diff
@@ -73,3 +78,30 @@ TEST_F(StateTreeTest, RegisterNewAddressLookupId) {
   EXPECT_OUTCOME_EQ(tree->registerNewAddress(address), kAddressId);
   EXPECT_OUTCOME_EQ(tree->lookupId(address), kAddressId);
 }
+
+/**
+ * walk visits hamt key-values
+ */
+TEST_F(StateTreeTest, Walk) {
+  using namespace fc;
```
Comment (ortyomka) [inline test/core/vm/state/state_tree_test.cpp:87]: Same

File: test/testutil/storage/ipld2.hpp [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include "storage/ipfs/datastore.hpp"
+#include "storage/ipld/ipld2.hpp"
+
+namespace fc {
+  struct IpldIpld2 : Ipld2 {
```
Comment (ortyomka) [inline test/testutil/storage/ipld2.hpp]: Maybe choose better name

File: test/core/storage/amt/amt_test.cpp (hunk 7 above)
Comment (ortyomka) [reply test/core/storage/amt/amt_test.cpp:161]: I mean "wrap all test cases"

**Review by Alexey-N-Chernyshov — APPROVED** [review]

**Pull request merged**